### PR TITLE
Add permission lifetime backend.

### PR DESCRIPTION
**Pull request by goodov**
Implement permission lifetime logic to track permissions lifetime and revoke them when required.

This change implements time-related aspect of brave/brave-browser#14126, i.e. it handles different lifetime options (24 hrs, 7 days, etc), but does not include support for "permission for this session" aka "revoke permission when all ```WebContents``` with eTLD+1 are closed".

The overall logic consists of these steps:
1. Observe UI interactions with ```PermissionPromptBubble``` to set permission lifetime in all ```PermissionRequest```s currently presented to a user.
2. Pass a permission decision to ```PermissionLifetimeManager``` (with all permission data, including content type, origins, lifetime).
3. Store permission origins and lifetime to prefs (```PermissionExpirations```).
4. Start/update permission revoke timer for a closest permission to be revoked (```PermissionLifetimeManager```).
5. Revoke permissions on timer hit and reschedule it for a next closest permission to revoke (```PermissionLifetimeManager```).
6. Observe ```HostContentSettingsMap``` to handle any permission changes not triggered via permission prompt dialogs (e.g. manually via Settings or via Page Info UI). This is required to properly update/stop the timer and clean up prefs for permissions we don't need to control anymore (```PermissionLifetimeManager```).

brave/brave-browser#14126

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [x] New files have MPL-2.0 license header
- [x] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [x] Changes in component dependencies are properly reflected in `gn`
- [x] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by simonhong** [comment]
@goodov Sorry for the late response. I'll review today!

**Comment by goodov** [comment]
PTAL. It's ready.

File: components/permissions/permission_lifetime_pref_names.h [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_PERMISSIONS_PERMISSION_LIFETIME_PREF_NAMES_H_
+#define BRAVE_COMPONENTS_PERMISSIONS_PERMISSION_LIFETIME_PREF_NAMES_H_
+
+namespace permissions {
+namespace prefs {
+
+extern const char kPermissionLifetimeRoot[];
```
Comment (simonhong) [inline components/permissions/permission_lifetime_pref_names.h:12]: nit: we recommend to use `constexpr` for string constants. With it, we don't need to create cc file.

**Review by simonhong — APPROVED** [review]
++ for me. Only have trivial nits. 👍

File: browser/permissions/BUILD.gn [hunk 2]
```diff
@@ -0,0 +1,63 @@
+source_set("permissions") {
+  check_includes = false
```
[the remark this answers is not in the document]
Comment (goodov) [reply browser/permissions/BUILD.gn:2]: I looked around and found ```sources.gni``` approach in some of our ```//brave/browser/<component>``` directories. Will redo it this way. My current approach is very hard to do without introducing a circular dep between ```//chrome/browser``` and ```//brave/browser```.

File: browser/permissions/permission_lifetime_manager_factory.cc [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+bool PermissionLifetimeManagerFactory::ServiceIsCreatedWithBrowserContext()
+    const {
+  return true;
+}
+
+content::BrowserContext*
+PermissionLifetimeManagerFactory::GetBrowserContextToUse(
+    content::BrowserContext* context) const {
+  return chrome::GetBrowserContextOwnInstanceInIncognito(context);
```
[the remark this answers is not in the document]
Comment (goodov) [reply browser/permissions/permission_lifetime_manager_factory.cc:60]: Yep, in incognito it has own instance to be in sync with own instance of ```HostContentSettingsMap```.

File: chromium_src/chrome/browser/ui/views/permission_bubble/permission_prompt_bubble_view.cc [hunk 4]
```diff
@@ -135,20 +135,32 @@ class PermissionLifetimeCombobox : public views::Combobox,
                << (lifetime ? lifetime->InSeconds() : -1);
     // TODO(https://github.com/brave/brave-browser/issues/14126): Set the
```
Comment (iefremov) [inline chromium_src/chrome/browser/ui/views/permission_bubble/permission_prompt_bubble_view.cc:136]: do you need the TODO ?

File: components/permissions/permission_expirations.h [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+      base::flat_map<ContentSettingsType, ExpiringPermissions>;
+
+  static void RegisterProfilePrefs(user_prefs::PrefRegistrySyncable* registry);
+
+  explicit PermissionExpirations(PrefService* prefs);
+  PermissionExpirations(const PermissionExpirations&) = delete;
+  PermissionExpirations& operator=(const PermissionExpirations&) = delete;
+  ~PermissionExpirations();
+
+  // Add expiring permission.
```
Comment (iefremov) [inline components/permissions/permission_expirations.h:43]: sounds descriptive :D

File: browser/permissions/permission_request_manager_browsertest.cc [hunk 6]
```diff
@@ -0,0 +1,847 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
```
Comment (iefremov) [inline browser/permissions/permission_request_manager_browsertest.cc:1]: do we really need this whole thing?

File: chromium_src/chrome/browser/ui/views/permission_bubble/permission_prompt_bubble_view.cc (hunk 4 above)
Comment (goodov) [reply chromium_src/chrome/browser/ui/views/permission_bubble/permission_prompt_bubble_view.cc:136]: no, this TODO is implemented in this PR. just forgot to remove the comment.

File: browser/permissions/permission_request_manager_browsertest.cc (hunk 6 above)
Comment (goodov) [reply browser/permissions/permission_request_manager_browsertest.cc:1]: we need test, we don't need chromium copyright.

Comment (goodov) [reply browser/permissions/permission_request_manager_browsertest.cc]: nope. removed.

File: components/permissions/DEPS [hunk 7]
```diff
@@ -1,3 +1,4 @@
 include_rules = [
   "+components/permissions",
+  "+services/preferences/public/cpp",
```
Comment (iefremov) [inline components/permissions/DEPS:3]: i gather we need slightly more deps here? `components/content_settings` etc.

File: chromium_src/components/permissions/permission_context_base.cc [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "components/permissions/permission_context_base.h"
+
+#define PermissionContextBase PermissionContextBase_ChromiumImpl
+#include "../../../../components/permissions/permission_context_base.cc"
+#undef PermissionContextBase
+
+#include "brave/components/permissions/permission_lifetime_manager.h"
```
Comment (iefremov) [inline chromium_src/components/permissions/permission_context_base.cc:12]: we need DEPS for `chromium_src` dirs as well

File: browser/permissions/sources.gni [hunk 9]
```diff
@@ -0,0 +1,6 @@
+brave_browser_permissions_sources = [
+  "//brave/browser/permissions/permission_lifetime_manager_factory.cc",
+  "//brave/browser/permissions/permission_lifetime_manager_factory.h",
+]
+
+brave_browser_permissions_deps = [ "//base" ]
```
Comment (bridiver) [inline browser/permissions/sources.gni:6]: you also need deps here for `//brave/components/permissions`, `components/permissions` and `components/keyed_service/content`

File: chromium_src/components/permissions/permission_context_base.cc (hunk 8 above)
Comment (bridiver) [reply chromium_src/components/permissions/permission_context_base.cc:12]: @iefremov we do have them so this must not be violating any existing rules

**Review by bridiver — CHANGES_REQUESTED** [review]
patches/chromium_src looks ok, but there are some missing dep issues

File: chromium_src/components/permissions/permission_context_base.cc (hunk 8 above)
Comment (goodov) [reply chromium_src/components/permissions/permission_context_base.cc:12]: @iefremov it's handled with this: https://github.com/brave/brave-core/blob/73f5e6f7d2ddd3051507f401dc55b134fb749b47/chromium_src/components/permissions/DEPS#L3

**Review by bridiver — APPROVED** [review]

**Pull request merged**